FWCore/Framework/src/SubProcess.cc

```diff
@@ -498,9 +508,54 @@ namespace edm {
     propagateProducts(InProcess, parentPrincipal, processBlockPrincipal);
 
     ProcessBlockTransitionInfo transitionInfo(processBlockPrincipal);
-    using Traits = OccurrenceTraits<ProcessBlockPrincipal, BranchActionGlobalEnd>;
-    endGlobalTransitionAsync<Traits>(
-        std::move(iHolder), *schedule_, transitionInfo, serviceToken_, subProcesses_, cleaningUpAfterException);
+
+    if (parentProducedProductIsKept(parentPrincipal, processBlockPrincipal)) {
+      auto runEndProcessBlock =
+          make_waiting_task([this, iWait = std::move(iHolder), info = transitionInfo, cleaningUpAfterException](
+                                std::exception_ptr const* iPtr) mutable {
+            ProcessBlockPrincipal& inputProcessBlockPrincipal = principalCache_.inputProcessBlockPrincipal();
+            inputProcessBlockPrincipal.clearPrincipal();
+            for (auto& s : subProcesses_) {
+              s.clearProcessBlockPrincipal(ProcessBlockType::Input);
+            }
+            if (iPtr) {
+              iWait.doneWaiting(*iPtr);
+            } else {
+              using Traits = OccurrenceTraits<ProcessBlockPrincipal, BranchActionGlobalEnd>;
+              endGlobalTransitionAsync<Traits>(
+                  std::move(iWait), *schedule_, info, serviceToken_, subProcesses_, cleaningUpAfterException);
+            }
+          });
+      WaitingTaskHolder holder(*iHolder.group(), runEndProcessBlock);
```

Comment:
Minor comment (or side note): I'd find 
```suggestion
      WaitingTaskHolder holder(*iHolder.group(), std::move(runEndProcessBlock));
```
here helpful to show that the ownership of `runEndProcessBlock` is transferred to the `holder`, but I suspect our `clang-tidy` would remove the `std::move` as unnecessary for raw pointer.

How bad would something along
```cpp
WaitingTaskHolder runEndProcessBlockHolder(*iHolder.group(), make_waiting_task(...));
```
look like in practice?

Reply:
This was a pattern of coding where I tried to copy something similar Chris did in another place.

I tried adding std::move and ran code-checks. scram b code-checks removed it as you suspected.

With your second suggestion, the entire content of the functor would be inside the WaitingTaskHolder constructor argument list. I am not sure that is any better... I don't have strong feelings about this one way or the other.

Reply:
> With your second suggestion, the entire content of the functor would be inside the WaitingTaskHolder constructor argument list. I am not sure that is any better...

That's exactly what I was afraid of, if it would obfuscate the code. The current way could well be the least bad.